rlbottraining/common_graders/rl_graders.py

```diff
@@ -29,13 +30,77 @@ def __init__(self, timeout_seconds=4.0, ally_team=0):
         ])
 
 class FailOnBallOnGroundAfterTimeout(FailOnTimeout):
+    def __init__(self, max_duration_seconds):
+        super().__init__(max_duration_seconds)
+        self.previous_ang_x = None
+        self.previous_ang_y = None
+        self.previous_ang_z = None
+        self.previous_total_goals = None
+
+    def set_previous_angular_velocity(self, ball, reset = False):
+        if not reset:
+            self.previous_ang_x = ball.angular_velocity.x
+            self.previous_ang_y = ball.angular_velocity.y
+            self.previous_ang_z = ball.angular_velocity.z
+            self.previous_vel_z = ball.velocity.z
+        else:
+            self.previous_ang_x = None
+            self.previous_ang_y = None
+            self.previous_ang_z = None
+            self.previous_vel_z = None
+
+    def set_previous_total_goals(self, total_goals, reset=False):
+        if not reset:
+            self.previous_total_goals = total_goals
+        else:
+            self.previous_total_goals = None
+
+    def current_total_goals(self, packet):
+        total_goals = 0
+        for car_id in range(packet.num_cars):
+            goal = packet.game_cars[car_id].score_info.goals
+            own_goal = packet.game_cars[car_id].score_info.own_goals
+            total_goals += goal + own_goal
+        return total_goals
 
     def on_tick(self, tick: TrainingTickPacket) -> Optional[Grade]:
         grade = super().on_tick(tick)
         if grade is None:
             return None
         assert isinstance(grade, FailOnTimeout.FailDueToTimeout)
         ball = tick.game_tick_packet.game_ball.physics
-        if ball.location.z < 100 and ball.velocity.z >= 0:
-            return grade
+        hit_ground = False
+
+        if self.previous_ang_z is None:
+            self.set_previous_angular_velocity(ball)
+        else:
+            max_ang_vel = 5.9999601985025075  #Max angular velocity possible
+            previous_ang_norm = math.sqrt(self.previous_ang_x**2 + self.previous_ang_y**2 + self.previous_ang_z**2)
+            if ball.location.z <= 1900:  #Making sure it doesnt count the ceiling
+                if (ball.angular_velocity.x != self.previous_ang_x or ball.angular_velocity.y != self.previous_ang_y):
+                    # If the ball hit anything its angular velocity will change in the x or y axis
+                    if self.previous_ang_z == ball.angular_velocity.z and not ( 130 < ball.location.z < 150) :
```

Comment:
This `130 < ball.location.z < 150` seems rather arbitrary. 
Could such a hit between the car/ball happen in other places? e.g. dribbling on top of the goal.
Could such a hit happen on that `z` on other maps?

Reply:
That is true that it could happens at another z, I did not like that way as well. After sleeping on the case I have some new ideas to try out, like seeing if the last touch was on this tick, and if the ball is a certain couple of units above one of the cars

Reply:
I think that'll make it better but you should keep some of the tricky cases in mind:

potential false positives:
- the ball is hit directly upwards using the nose of the car
- the punching glove mutator hitting the ball upwards.

potential false negatives:
- the ball bounces off the flat top of the goal while the car is directly under the ceiling of the goal.

Reply:
potential false positives:

> - the ball is hit directly upwards using the nose of the car
seeing the the last touch was this tick should avoid this problem
> - the punching glove mutator hitting the ball upwards.
punching doesn't update the last touch. this may become a problem, but currently we cant test it

potential false negatives:
> - the ball bounces off the flat top of the goal while the car is directly under the ceiling of the goal.
Once again, the last touch check should not allow this to happen

Reply:
:+1: I didn't think of using the last touch info.